source/server/listener_impl.cc

```diff
@@ -386,45 +383,47 @@ void ListenerImpl::buildAccessLog() {
-
-    udp_listener_worker_router_ =
-        std::make_unique<Network::UdpListenerWorkerRouterImpl>(concurrency);
-  }
-}
-
-void ListenerImpl::buildUdpWriterFactory(Network::Socket::Type socket_type) {
-  if (socket_type == Network::Socket::Type::Datagram) {
-    auto udp_writer_config = config_.udp_writer_config();
-    if (!Api::OsSysCallsSingleton::get().supportsUdpGso() ||
-        udp_writer_config.typed_config().type_url().empty()) {
-      const std::string default_type_url =
-          "type.googleapis.com/envoy.config.listener.v3.UdpDefaultWriterOptions";
-      udp_writer_config.mutable_typed_config()->set_type_url(default_type_url);
-    }
-    auto& config_factory =
-        Config::Utility::getAndCheckFactory<Network::UdpPacketWriterConfigFactory>(
-            udp_writer_config);
-    ProtobufTypes::MessagePtr message = Config::Utility::translateAnyToFactoryConfig(
-        udp_writer_config.typed_config(), validation_visitor_, config_factory);
-    udp_writer_factory_ = config_factory.createUdpPacketWriterFactory(*message);
+  if (socket_type != Network::Socket::Type::Datagram) {
+    return;
   }
+  if (!config_.reuse_port() && concurrency > 1) {
+    throw EnvoyException("Listening on UDP when concurrency is > 1 without the SO_REUSEPORT "
+                         "socket option results in "
+                         "unstable packet proxying. Configure the reuse_port listener option or "
+                         "set concurrency = 1.");
+  }
+  auto udp_config = config_.udp_listener_config();
+  if (udp_config.listener_config().typed_config().type_url().empty()) {
+    const std::string default_type_url =
+        "type.googleapis.com/envoy.config.listener.v3.ActiveRawUdpListenerConfig";
+    udp_config.mutable_listener_config()->mutable_typed_config()->set_type_url(default_type_url);
+  }
+
+  auto& listener_config_factory =
+      Config::Utility::getAndCheckFactory<ActiveUdpListenerConfigFactory>(
+          udp_config.listener_config());
+  ProtobufTypes::MessagePtr listener_typed_config = Config::Utility::translateAnyToFactoryConfig(
+      udp_config.listener_config().typed_config(), validation_visitor_, listener_config_factory);
+
+  udp_listener_config_ = std::make_unique<UdpListenerConfigImpl>(udp_config);
+  udp_listener_config_->listener_factory_ =
+      listener_config_factory.createActiveUdpListenerFactory(*listener_typed_config, concurrency);
+  udp_listener_config_->listener_worker_router_ =
+      std::make_unique<Network::UdpListenerWorkerRouterImpl>(concurrency);
+
+  auto writer_config = udp_config.writer_config();
+  if (!Api::OsSysCallsSingleton::get().supportsUdpGso() ||
+      writer_config.typed_config().type_url().empty()) {
+    const std::string default_type_url =
+        "type.googleapis.com/envoy.config.listener.v3.UdpDefaultWriterOptions";
+    writer_config.mutable_typed_config()->set_type_url(default_type_url);
+  }
```

Comment:
if it does support GRO do we need to do size bounds checks to make sure folks don't configure it too high?

Reply:
Sorry configure what? I added PGV annotations to max out at 64K.